src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProject.BatchingDocumentCollection.cs

```diff
@@ -528,6 +529,11 @@ public void ReorderFiles(ImmutableArray<string> filePaths)
             }
         }
 
+        /// <summary>
+        /// Updates the solution for a set of batch changes.
+        /// While it is OK for this method to *read* local state, it cannot *modify* it as this may
+        /// be called multiple times (when the workspace update fails due to interceding updates).
+        /// </summary>
```

Comment:
note; one to *possibly* solve this (though i don't know all the data involved here) is to call into a static helper, passing the local state, but not the `this` pointer.  This would prevent doing things like assigning to field-instance-state.

it wouldn't prevent mutation of the values if they are mutable though.  it might also be really ugly depending on how much instance data you'd have to pass to the static method.